### ISSUE TITLE
Funnel orderings

### ISSUE THREAD

**Issue by alexkim205** [PostHog/posthog#5552]
## Context

This funnel ordering feature provides more granularity for users are more concerned with the steps getting completed, rather than in which order they were completed.

I separated this out into its own issue so that discussion about this specific feature can happen in its own thread. Feel free to add any more relevant information that I might've missed!

- [Nail Funnel Ordering Section Specs](https://docs.google.com/document/d/1S7fwha7cph-79YaPADfFKHILCnPmBgk4rqaMu-gPZ3g/edit#heading=h.nwp8ynw69xae)
- Master Ticket - #5367 
- Backend Specs - [Comment](https://github.com/PostHog/posthog/issues/5367#issuecomment-889860866)
   - FunnelOrderType: `STRICT | ORDERED | UNORDERED`

Figma [mockup](https://www.figma.com/file/1quiOoOkygRZO4h8gRxegZ/%5BPublic%5D---Nail-Funnels-Prototype?node-id=0%3A1)

<img width="336" alt="Screen Shot 2021-08-11 at 1 44 41 PM" src="https://user-images.githubusercontent.com/13460330/129100855-caf3f24d-d8de-4a9a-aeb0-ad942a3cff9c.png">

## A few questions

I think @paolodamico @clarkus have the most context here so CC'ing ya'll for guidance. 

1. Ordered and unordered order types seem pretty self explanatory but what does the strict type mean?
2. Will this be available for steps, trends, and conversion graph types? Will the intent of ordering be clear when the steps aren't being shown (in conversion and trends)?
3. Do we have mockups for “any order” funnels that are ready to go? If not we can also use this thread for brainstorming
    a. Will the user use the same funnel filter tab for both unordered and ordered steps?
4. The copy in the mockup above makes it sound like the steps can be reordered arbitrarily and I can't really wrap my head around what this means. Does this mean unordered steps will catch all events that satisfy any of the filter steps or does it mean that just the steps will be visually reordered with the same data?

**Comment by paolodamico** [comment]
TLDR; probably worth waiting for Chris to be back to have some input on this

1. I'll let @neilkakkar or @EDsCODE confirm but I believe strict means that no other steps are allowed between each step. I'm not sure if we should implement this yet, maybe we can implement and test usage? In this case probably a dropdown makes more sense, and we may want to change the visuals on the glyphs too to indicate this. 
2. Again @neilkakkar can you confirm capabilities-wise? I think the intent of the ordering can be clear, just by implementing a visual change like suggested above.
3. I think we can wait for @clarkus to confirm on mockups but I would think we can use the same for ordered and unordered steps.
4. Yes, technically steps could be reordered arbitrarily, I don't think it's problematic in terms of how the funnel is constructed, but how it's displayed. One thing in particular that I have a hard time wrapping my head around is conversion/drop-off rates between steps.

**Comment by neilkakkar** [comment]
1. What Paolo said! It means no events are allowed between the events we want to see in the funnel.
2. Trends and Time-to-convert both have all orderings. 

\4. To add: the backend semantics for unordered funnels mean: The first item in array is "number of users who did atleast 1 event from the funnel", the second is "number of users who did atleast 2 events from the funnel".. and so on. There's no concept here of which event came first. 
So, I'd suggest incorporating this into the UI, and not mentioning the specific step at all, which dissolves the idea of reordering, since there's no specific step sequence to reorder in.

**Comment by clarkus** [comment]
There aren't any mockups for this other than the one shared above. I agree with @neilkakkar's point regarding item 4. I can work to clarify that today and post back here. Thanks for breaking down the problem into an easy-to-understand way.

**Comment by clarkus** [comment]
I have a couple of solutions in mind for this, but I wanted to clarify some things first to ensure I'm understanding correctly. 

We allow for 3 evaluation options:

1) **Sequential** - these steps must happen in this sequence, but other events can happen in between steps. For example, step 2 must occur after step 1, but it's OK if a bunch of other stuff happened in between those events. 
2) **Strict** - these steps must happen in this sequence, and only in this sequence. For example, persons must enter step 2 from step 1 and only from step 1. 
3) **Any** - persons must complete all of these steps, but they can occur in any sequence. For example, a person can complete a 4 step funnel in any sequence (step 4 > step 2 > step 1 > step 3)

Do we allow for any combination of these options? How granular should the control be for this? Can I make any step "strict" while others are able to occur in any sequence? For example allowing a range of steps earlier in the funnel but requiring the last two steps be strict in their evaluation - (•) pageview, (•) view movie, (•) watch trailer, (4) add to cart, (5) purchase.

**Comment by clarkus** [comment]
This includes work from #3408, but here's an idea for a global control to set the funnel evaluation parameter. The options are sequential order, strict order, or any order. Tooltips are shown for actions... there's more here than necessary for this issue. I've written some rough descriptions for each option... please feel free to help me refine that language to be more accurate / approachable. 

![Funnel steps - Step Order](https://user-images.githubusercontent.com/254612/129640692-661040c7-ad51-4803-bd34-757ba8f2dbb7.png)
![Funnel steps - Step Order-1](https://user-images.githubusercontent.com/254612/129640693-773fa5bb-93b9-453c-9da8-c53f488ff328.png)

**Comment by alexkim205** [comment]
This looks great! Thank you so much

1. Those three evaluation options are correct 
2. As of now, a combination of these options isn't available. I think the dropdown you have in the mockup should work great and I like that its location inside the steps card indicates that its used to modify a steps parameter. Having a combination of options might be a feature request in the future, but I feel like conveying that in a concise non-cluttered way might be tricky.


**Some scattered feedback**

- The copy here is clear to me. Maybe specifying Step 1 and Step 2 in the copy seems like its referring to specific steps to the left. Maybe Step A and Step B would work w/o sacrificing this ordering concept? What do you think?
- If the user selects "Any order" should we hide the sequential step numbers?
- How does renaming work here? Will capturing this input happen in a modal or inside the input field of the step? I don't think it's possible to edit funnel steps yet and have them persist across browser sessions/in backend cache, but definitely would be a helpful feature to have.

**Comment by clarkus** [comment]
I think it's fine to leave numbers for the step identifiers. They're still specified in a sequence even if evaluation is variable. I could see the user specifying a few steps and pivoting between step evaluation options. Even in the "any order" case I don't think we need to change the step identifiers. If we ever allow step-level evaluation rules, then I could see distinguishing things more via alternate step IDs. 

The renaming is documented at https://github.com/PostHog/posthog/issues/5229.
